### PR TITLE
Refactoring bestMatching() score computation

### PR DESCRIPTION
**Pull request by mccartney**
Pure refactoring, no change in the behavior.
Refactored the `bestMatching()` function for picking the right constructor by broadening the definition of the ad-hoc `Score` notion. The optionals and defaults have been incorporated into `Score`.

Added quite a few unit tests for this code which were missing.

### CODE REVIEW

File: tests/src/test/scala/org/json4s/reflect/DescriptorsSpec.scala [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+      val descriptor: ClassDescriptor = describe(classOf[Citizen])
+
+      // when
+      val best = descriptor.bestMatching(List("firstName", "lastName", "newInfoWhichDoesNotExist")).get.constructor
+
+      // test
+      best.getParameterTypes() must_== Array(classOf[String], classOf[String], classOf[String])
+    }
+
+  }
```
Comment (mccartney) [inline tests/src/test/scala/org/json4s/reflect/DescriptorsSpec.scala:112]:
```suggestion
  }
```

File: tests/src/test/scala/org/json4s/reflect/DescriptorsSpec.scala [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  object Person {
+    def apply(email: String) = new Person("Russell", "Westbrook")
+  }
+
+  case class Citizen(firstName: String, lastName: String, id: String) {
+    def this(firstName: String, lastName: String, idAsANumber: Int) = this(firstName, lastName, idAsANumber.toString)
+  }
+
+  case class Company(name: String, industry: String = "IT", ceo: Person = new Person("My", "Self"), yearFounded: Int) {
+    def this(name: String, industry: String, ceo: Person) {
```
Comment (magnolia-k) [inline tests/src/test/scala/org/json4s/reflect/DescriptorsSpec.scala:19]:
```scala
def this(name: String, industry: String, ceo: Person) = {
```

Shouldn't there be an '=' between the method definition and the code block?
(2.13.3 will cause a compile error)

Comment (mccartney) [reply tests/src/test/scala/org/json4s/reflect/DescriptorsSpec.scala:19]: I think this is what Travis build says too.

**Comment by seratch** [comment]
After merging this, the CI builds started failing on the `master` branch due to some reason. So, we've reverted it by #700.

@mccartney Sorry for bothering you but could you check this and come up with another PR?

**Comment by mccartney** [comment]
Sure. I wonder why Travis CI didn't reflect that before the merging.

**Pull request merged**